### DISCUSSION TITLE
Speed-up sonic indexing of an existing archive

### DISCUSSION THREAD

**Discussion by mwnoo** [ArchiveBox/ArchiveBox#1596] [Q&A] [answered]
I updated my setup to Archivebox 0.7.2 and want to rebuild the sonic index from scratch. My index contains about 7k snapshots. I removed the sonic folder `kv` and `fst` from my archive, rebooted and ran: `archivebox update --index-only`

After a couple of minutes I see that tmp files are created in the archive folder and that urls are shown in the terminal. But the flow of urls seems to freeze at some point. In the sonic `kv` folder there are still `*.sst` files created. So I think sonic is still working on the index.

After a day also new urls are displayed in the terminal and again the urls seem to freeze at some point while sonic still seems to be working (activity in the `kv` folder).

- Is it normal that rebuilding the sonic index from scratch takes such a long time? Is is possible to speed-up this process?
- Are 7k snapshots too much for sonic to index? (I've seen people in the forum running sonic on archives containing 20k snapshots)

I also copied my setup to a machine with much more resources (20 cores, 16 GB Ram, Nvme disks) but still indexing is rather slow.  I don't see a really improvement in speed. 
- Is it save to copy the sonic folders (`kv` and `fst`) from one setup to another given that the used archive is the same?

**Comment by pirate** [comment, marked as answer]
Sonic itself should definitely not be that slow, I am using it with 100k+ snapshots on my main archive, it can easily handle 1M+ records in testing on a 1CPU machine with NVMe disks.

It's `archivebox update` itself that is slow, it's terribly optimizied in 0.7.2 (it loads all snapshots into memory many times over to do deduping and other checks). My apologies, the usual use-case is running sonic from the beginning to index as you go, and `archivebox update --index-only` was added as an afterthought.

I'm working on speeding it up massively in >= v0.8.5!

> Is it safe to copy the sonic folders (kv and fst) from one setup to another

Yes it should be afaik

>  Is is possible to speed-up this process?

For now you can temporarily work around the slowness by only indexing a smaller range at a time.
Use `archivebox update --index-only --after <xyz> --before <xyz>`, and do chunks of 2000 URLs at a time. No need for a powerful machine as long as you limit the total snapshots handled at a time.

Use `archivebox update --help` to see all the options for selecting a range.

**Reply by mwnoo** [reply]
Ok, that could explain why creating the index is so slow.

I restored an older version of my archive running v0.6.2. Creating the index seems to run a lot faster (i.e. faster increment of *.sst files in sonic folder; now working on 130000.sst ). However, `archivebox update --index-only` is again running 24+ hours now and still not finished with 7k snapshots.

**Reply by pirate** [reply]
`<=v0.8.6` did some ridiculous `O(n^2)` loops where it loads the *entire index* into memory when processing *each snapshot* when updating. I definitely advise using the `--after <xyz> --before <xyz>` as the time-to-run scales quadratically to infinity with number of snapshots on the old versions!

**Reply by mwnoo** [reply]
I tried it with the `--after` and `--before` arguments. This works fine for a handful of snapshots but not for more than 500+. 
Maybe I need to craft a loop that updates at every single timestamp to prevent this issue or probably better to wait for the v.0.8.x release to recreate my index. Thanks, priate!
